File: poly_apps/laravel_main/app/Services/AiGateway/AiImageHistory.php
<?php

namespace App\Services\AiGateway;

use App\Providers\PathMapper;

/**
 * Shared AI image-generation history — the PHP twin of pycore's
 * pyctl.ai.ai_image_history, sharing the SAME on-disk store.
 *
 * pycore (Windows host) and Laravel (WSL) read/write ONE set of files so the
 * image-tools gallery is identical on both ends. The only filesystem path both
 * resolve to a single physical file is the core_node repo root (D:\..\core_node
 * on Windows == /mnt/d/..\core_node in WSL via DrvFs), so the store lives under
 * <core_node>/.ai_state — the same channel ai_rate_usage.json / .secret_keys use:
 *
 *   <core_node>/.ai_state/ai_image_history.json   newest-LAST index (ring 200)
 *   <core_node>/.ai_state/ai_images/<id>.<ext>    the generated image bytes
 *
 * Index doc:  { version: 1, saved_at: float, entries: [ <entry>, ... ] }
 * Entry (NO base64 in the index — bytes live in the image file):
 *   { id, ts, iso, provider, model, prompt, size, mime, bytes, file,
 *     latency_ms, source, origin: 'pycore'|'laravel', ok }
 *
 * Write safety mirrors AiRateLimiter: a tmp file + atomic rename, serialized by
 * an flock'd lock file so concurrent Octane workers (and the pycore atomic
 * replace) never corrupt the document.
 */
class AiImageHistory
{
    /** Newest-last ring buffer cap; older entries + their image files are trimmed. */
    private const MAX_ENTRIES = 200;

    /** Bound prompt length stored in the index (matches pycore). */
    private const PROMPT_MAX = 2000;

    /** mime -> file extension (matches pycore _MIME_EXT, default png). */
    private const MIME_EXT = [
        'image/png' => 'png',
        'image/jpeg' => 'jpg',
        'image/jpg' => 'jpg',
        'image/webp' => 'webp',
        'image/gif' => 'gif',
    ];

    /** Absolute path of the shared .ai_state dir (same dir pycore writes). */
    public static function stateDir(): string
    {
        $coreNode = PathMapper::getCoreNodeDir() ?: PathMapper::getLaravelMainDir();
        return rtrim($coreNode, '/\\') . '/.ai_state';
    }

    /** Absolute path of the shared history index file. */
    public static function indexFile(): string
    {
        return self::stateDir() . '/ai_image_history.json';
    }

    /** Absolute path of the shared image bytes dir. */
    public static function imagesDir(): string
    {
        return self::stateDir() . '/ai_images';
    }

    /**
     * Persist a generated image (bytes file + index entry) to the shared store.
     *
     * Best-effort: any storage failure returns null (never throws) so image
     * delivery to the caller never depends on history succeeding.
     *
     * @param  array<string, mixed>  $meta  provider, model, prompt, size, mime, latency_ms, source, origin, ok
     * @return array<string, mixed>|null    the stored index entry (NO base64), or null
     */
    public static function record(array $meta, string $imageBase64): ?array
    {
        if ($imageBase64 === '') {
            return null;
        }
        $raw = base64_decode($imageBase64, true);
        if ($raw === false || $raw === '') {
            return null;
        }

        $ts = microtime(true);
        $provider = (string) ($meta['provider'] ?? '');
        $prompt = (string) ($meta['prompt'] ?? '');
        $mime = (string) ($meta['mime'] ?? 'image/png');
        if ($mime === '') {
            $mime = 'image/png';
        }
        // Match pycore: sha1("<ts>:<provider>:<prompt>"), first 16 hex chars.
        $id = substr(sha1($ts . ':' . $provider . ':' . $prompt), 0, 16);
        $ext = self::extFor($mime);
        $rel = 'ai_images/' . $id . '.' . $ext;

        $latency = $meta['latency_ms'] ?? null;
        $entry = [
            'id' => $id,
            'ts' => $ts,
            // Match pycore datetime.fromtimestamp(ts, utc).isoformat(timespec="seconds"),
            // which renders the offset as "+00:00" (not "Z").
            'iso' => gmdate('Y-m-d\TH:i:s+00:00', (int) $ts),
            'provider' => $provider,
            'model' => (string) ($meta['model'] ?? ''),
            'prompt' => self::clipPrompt($prompt),
            'size' => (string) ($meta['size'] ?? ''),
            'mime' => $mime,
            'bytes' => strlen($raw),
            'file' => $rel,
            'latency_ms' => $latency === null ? null : (is_numeric($latency) ? $latency + 0 : $latency),
            'source' => (string) ($meta['source'] ?? 'image'),
            'origin' => (string) ($meta['origin'] ?? 'laravel'),
            'ok' => array_key_exists('ok', $meta) ? (bool) $meta['ok'] : true,
        ];

        return self::withLock(static function () use ($entry, $rel, $raw): ?array {
            $imagesDir = self::imagesDir();
            if (!is_dir($imagesDir)) {
                @mkdir($imagesDir, 0775, true);
            }
            if (@file_put_contents(self::stateDir() . '/' . $rel, $raw) === false) {
                return null;
            }
            $doc = self::load();
            $doc['entries'][] = $entry;
            self::trim($doc);
            self::save($doc);
            return $entry;
        });
    }

    /**
     * Newest-first index entries (metadata only), capped at $limit (1..200).
     *
     * @return array<int, array<string, mixed>>
     */
    public static function list(int $limit = 50): array
    {
        $limit = max(1, min(self::MAX_ENTRIES, $limit));
        $entries = self::load()['entries'];
        $entries = array_reverse($entries);
        return array_slice($entries, 0, $limit);
    }

    /**
     * Image bytes + mime for a stored id, or null when missing.
     *
     * @return array{bytes: string, mime: string}|null
     */
    public static function read(string $id): ?array
    {
        $id = trim($id);
        if ($id === '') {
            return null;
        }
        $rel = null;
        $mime = 'image/png';
        foreach (self::load()['entries'] as $e) {
            if (($e['id'] ?? null) === $id) {
                $rel = $e['file'] ?? null;
                $mime = (string) ($e['mime'] ?? 'image/png');
                break;
            }
        }
        if ($rel === null) {
            return null;
        }
        $bytes = @file_get_contents(self::stateDir() . '/' . $rel);
        if ($bytes === false) {
            return null;
        }
        return ['bytes' => $bytes, 'mime' => $mime !== '' ? $mime : 'image/png'];
    }

    /** Remove one history entry + its image file. True when an entry was removed. */
    public static function delete(string $id): bool
    {
        $id = trim($id);
        if ($id === '') {
            return false;
        }
        return (bool) self::withLock(static function () use ($id): bool {
            $doc = self::load();
            $keep = [];
            $removed = null;
            foreach ($doc['entries'] as $e) {
                if ($removed === null && ($e['id'] ?? null) === $id) {
                    $removed = $e;
                } else {
                    $keep[] = $e;
                }
            }
            if ($removed === null) {
                return false;
            }
            $doc['entries'] = $keep;
            self::save($doc);
            $rel = $removed['file'] ?? null;
            if ($rel) {
                @unlink(self::stateDir() . '/' . $rel);
            }
            return true;
        });
    }

    /** Delete ALL history entries + image files. Returns the count removed. */
    public static function clear(): int
    {
        return (int) self::withLock(static function (): int {
            $doc = self::load();
            $base = self::stateDir();
            foreach ($doc['entries'] as $e) {
                $rel = $e['file'] ?? null;
                if ($rel) {
                    @unlink($base . '/' . $rel);
                }
            }
            $n = count($doc['entries']);
            $doc['entries'] = [];
            self::save($doc);
            return $n;
        });
    }

    // --- internals --------------------------------------------------------- //

    private static function extFor(string $mime): string
    {
        $key = strtolower(trim(explode(';', $mime)[0]));
        return self::MIME_EXT[$key] ?? 'png';
    }

    private static function clipPrompt(string $prompt): string
    {
        // pycore slices the first 2000 chars (unicode code points). mb_substr
        // matches that; ASCII prompts are identical either way.
        return mb_strlen($prompt) > self::PROMPT_MAX ? mb_substr($prompt, 0, self::PROMPT_MAX) : $prompt;
    }

    /**
     * Load the shared index doc, normalizing to { version, saved_at, entries }.
     * A corrupt/missing index yields a fresh empty doc (never throws).
     *
     * @return array{version: int, saved_at: float, entries: array<int, array<string, mixed>>}
     */
    private static function load(): array
    {
        $path = self::indexFile();
        if (is_file($path)) {
            $raw = @file_get_contents($path);
            if ($raw !== false && $raw !== '') {
                $decoded = json_decode($raw, true);
                if (is_array($decoded) && isset($decoded['entries']) && is_array($decoded['entries'])) {
                    $decoded['version'] = isset($decoded['version']) ? (int) $decoded['version'] : 1;
                    $decoded['saved_at'] = isset($decoded['saved_at']) ? (float) $decoded['saved_at'] : 0.0;
                    return $decoded;
                }
            }
        }
        return ['version' => 1, 'saved_at' => 0.0, 'entries' => []];
    }

    /**
     * Trim to the newest MAX_ENTRIES, deleting dropped entries' image files.
     *
     * @param  array{entries: array<int, array<string, mixed>>}  $doc
     */
    private static function trim(array &$doc): void
    {
        $entries = $doc['entries'];
        $count = count($entries);
        if ($count <= self::MAX_ENTRIES) {
            return;
        }
        $cut = $count - self::MAX_ENTRIES;
        $drop = array_slice($entries, 0, $cut);
        $doc['entries'] = array_values(array_slice($entries, $cut));
        $base = self::stateDir();
        foreach ($drop as $e) {
            $rel = $e['file'] ?? null;
            if ($rel) {
                @unlink($base . '/' . $rel);
            }
        }
    }

    /**
     * Atomically write the index doc (tmp file + rename), matching pycore's
     * os.replace and AiRateLimiter::save. version stays 1, saved_at refreshed.
     *
     * @param  array{version?: int, saved_at?: float, entries: array<int, mixed>}  $doc
     */
    private static function save(array $doc): void
    {
        $path = self::indexFile();
        $dir = dirname($path);
        if (!is_dir($dir)) {
            @mkdir($dir, 0775, true);
        }
        $doc['version'] = 1;
        $doc['saved_at'] = microtime(true);
        $tmp = $path . '.tmp.' . getmypid();
        // ensure_ascii=False parity: keep unicode + slashes unescaped.
        if (@file_put_contents($tmp, json_encode($doc, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE)) !== false) {
            @rename($tmp, $path);
        } else {
            @unlink($tmp);
        }
    }

    /**
     * Serialize a read-modify-write across Octane workers with an flock'd lock
     * file (same pattern as AiRateLimiter). Cross-runtime (vs pycore) safety
     * still rests on the atomic rename in save(); image generations are seconds
     * apart so the lost-update window is negligible.
     *
     * @template T
     * @param  callable():T  $fn
     * @return T
     */
    private static function withLock(callable $fn)
    {
        $dir = self::stateDir();
        if (!is_dir($dir)) {
            @mkdir($dir, 0775, true);
        }
        $lockPath = self::indexFile() . '.lock';
        $handle = @fopen($lockPath, 'c');
        if ($handle === false) {
            return $fn();
        }
        $locked = @flock($handle, LOCK_EX);
        try {
            return $fn();
        } finally {
            if ($locked) {
                @flock($handle, LOCK_UN);
            }
            @fclose($handle);
        }
    }
}

File: pycore/pyctl/ai/ai_image_history.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Shared AI image-generation history (cross-runtime: pycore on Windows + Laravel
in WSL share ONE file, exactly like the rate-usage store).

Both runtimes read/write the SAME files under ``<core_node>/.ai_state`` — the
only filesystem location both see as a single file (D:\\..\\core_node on Windows
== /mnt/d/..\\core_node in WSL via DrvFs), the same channel ``.secret_keys`` and
``ai_rate_usage.json`` already use. The Laravel side mirrors this contract in
``App\\Services\\AiGateway\\AiImageHistory``.

Layout:
  <core_node>/.ai_state/ai_image_history.json   — newest-last index (ring buffer)
  <core_node>/.ai_state/ai_images/<id>.<ext>    — the generated image bytes

Index entry (the only shape the UI / Laravel depend on; NO base64 in the index):
  { id, ts, iso, provider, model, prompt, size, mime, bytes, file,
    latency_ms, source, origin: 'pycore'|'laravel', ok }

Safety mirrors ai_rate_limits: tmp file + atomic os.replace + an in-process lock.
Cross-runtime DrvFs locking is unreliable, but image generations are seconds+
apart so the lost-update window is negligible and atomic replace prevents
corruption. All imports at file top (PYTHON_PYCORE.md §1.4); ColorPrint logging.
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
import threading
import time
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Tuple

from pycore.pyfoundations.pybasecommon.color_print import ColorPrint
from pycore.pyfoundations.system_paths import APP_DATA_DIR, get_core_node_root, get_local_data_dir

# Cross-runtime shared store (see module docstring + ai_rate_limits rationale).
# Lives under <cache>/pycore/.ai_state; the prior <core_node>/.ai_state location
# is migrated once on first access.
_SHARED_STATE_DIR = get_local_data_dir() / ".ai_state"
_OLD_SHARED_DIR = get_core_node_root() / ".ai_state"
_LEGACY_DIR = APP_DATA_DIR / "ai_state"

# Newest-last ring buffer cap; older entries (and their image files) are trimmed.
_MAX_ENTRIES = 200

_MIME_EXT = {
    "image/png": "png", "image/jpeg": "jpg", "image/jpg": "jpg",
    "image/webp": "webp", "image/gif": "gif",
}

_lock = threading.Lock()


def _migrate_old_state():
    """Move files from the prior shared dir (<core_node>/.ai_state) into the new
    <cache>/pycore/.ai_state once. Idempotent — a name already present in the
    new dir is left alone. All three AI-state modules share this dir, so whichever
    runs first relocates everything (rate usage, usage records, image history +
    the ai_images/ folder)."""
    try:
        if not _OLD_SHARED_DIR.exists() or _OLD_SHARED_DIR.resolve() == _SHARED_STATE_DIR.resolve():
            return
        for item in _OLD_SHARED_DIR.iterdir():
            dest = _SHARED_STATE_DIR / item.name
            if not dest.exists():
                os.replace(str(item), str(dest))
    except Exception:
        pass


def _state_dir():
    """Shared ``<cache>/pycore/.ai_state`` dir (legacy fallback if
    the cache root is not writable, e.g. a read-only deploy)."""
    try:
        _SHARED_STATE_DIR.mkdir(parents=True, exist_ok=True)
        _migrate_old_state()
        return _SHARED_STATE_DIR
    except Exception:
        _LEGACY_DIR.mkdir(parents=True, exist_ok=True)
        return _LEGACY_DIR


def _index_file():
    return _state_dir() / "ai_image_history.json"


def _images_dir():
    d = _state_dir() / "ai_images"
    try:
        d.mkdir(parents=True, exist_ok=True)
    except Exception:
        pass
    return d


def _load_index() -> Dict[str, Any]:
    path = _index_file()
    try:
        if path.is_file():
            data = json.loads(path.read_text(encoding="utf-8"))
            if isinstance(data, dict) and isinstance(data.get("entries"), list):
                return data
    except Exception as e:  # noqa: BLE001 — a corrupt index must never crash callers
        ColorPrint.yellow(f"[ai_image_history] index unreadable ({e}); starting fresh")
    return {"version": 1, "saved_at": 0.0, "entries": []}


def _save_index(doc: Dict[str, Any]) -> None:
    doc["saved_at"] = time.time()
    path = _index_file()
    tmp = path.with_suffix(".json.tmp")
    try:
        tmp.write_text(json.dumps(doc, ensure_ascii=False), encoding="utf-8")
        os.replace(tmp, path)
    except Exception as e:  # noqa: BLE001
        ColorPrint.yellow(f"[ai_image_history] index write failed: {e}")
        try:
            if tmp.exists():
                tmp.unlink()
        except OSError:
            pass


def _ext_for(mime: str) -> str:
    return _MIME_EXT.get((mime or "").lower().split(";")[0].strip(), "png")


def _trim(doc: Dict[str, Any]) -> None:
    """Keep only the newest _MAX_ENTRIES; delete dropped entries' image files."""
    entries = doc.get("entries") or []
    if len(entries) <= _MAX_ENTRIES:
        return
    drop = entries[: len(entries) - _MAX_ENTRIES]
    doc["entries"] = entries[len(entries) - _MAX_ENTRIES:]
    base = _state_dir()
    for e in drop:
        rel = e.get("file")
        if rel:
            try:
                (base / rel).unlink()
            except OSError:
                pass


def record_image(
    *,
    provider: str,
    model: str,
    prompt: str,
    image_base64: str,
    size: Optional[str] = None,
    mime: str = "image/png",
    latency_ms: Optional[float] = None,
    source: str = "image",
    origin: str = "pycore",
    ok: bool = True,
) -> Optional[Dict[str, Any]]:
    """Persist a generated image (file + index entry) to the shared store.

    Returns the index entry (NO base64) or None when there is nothing to store.
    Best-effort: a storage failure is logged and swallowed, never raised — image
    delivery to the caller must not depend on history succeeding.
    """
    if not image_base64:
        return None
    try:
        raw = base64.b64decode(image_base64)
    except Exception:  # noqa: BLE001
        return None
    if not raw:
        return None

    ts = time.time()
    digest = hashlib.sha1(f"{ts}:{provider}:{prompt}".encode("utf-8")).hexdigest()[:16]
    ext = _ext_for(mime)
    rel = f"ai_images/{digest}.{ext}"
    entry = {
        "id": digest,
        "ts": ts,
        "iso": datetime.fromtimestamp(ts, timezone.utc).isoformat(timespec="seconds"),
        "provider": provider or "",
        "model": model or "",
        "prompt": (prompt or "")[:2000],
        "size": size or "",
        "mime": mime or "image/png",
        "bytes": len(raw),
        "file": rel,
        "latency_ms": latency_ms,
        "source": source or "image",
        "origin": origin or "pycore",
        "ok": bool(ok),
    }
    with _lock:
        try:
            (_images_dir().parent / rel).write_bytes(raw)
        except Exception as e:  # noqa: BLE001
            ColorPrint.yellow(f"[ai_image_history] image write failed: {e}")
            return None
        doc = _load_index()
        doc["entries"].append(entry)
        _trim(doc)
        _save_index(doc)
    ColorPrint.green(
        f"[ai_image_history] recorded {provider}/{model} ({len(raw)//1024}KB) id={digest}")
    return entry


def list_history(limit: int = 50) -> List[Dict[str, Any]]:
    """Newest-first index entries (metadata only), capped at ``limit`` (1..200)."""
    limit = max(1, min(_MAX_ENTRIES, int(limit) if str(limit).isdigit() else 50))
    with _lock:
        entries = list((_load_index().get("entries") or []))
    return list(reversed(entries))[:limit]


def read_image(image_id: str) -> Tuple[bytes, str]:
    """(bytes, mime) for a stored image id, or (b'', '') when missing."""
    image_id = (image_id or "").strip()
    if not image_id:
        return b"", ""
    with _lock:
        for e in (_load_index().get("entries") or []):
            if e.get("id") == image_id:
                rel = e.get("file")
                mime = e.get("mime") or "image/png"
                break
        else:
            return b"", ""
    try:
        return (_state_dir() / rel).read_bytes(), mime
    except Exception:  # noqa: BLE001
        return b"", ""


def entry_path(image_id: str) -> Optional[str]:
    """Absolute path of a stored image file (for reveal / show-location), or None."""
    image_id = (image_id or "").strip()
    if not image_id:
        return None
    with _lock:
        for e in (_load_index().get("entries") or []):
            if e.get("id") == image_id and e.get("file"):
                try:
                    return str((_state_dir() / e["file"]).resolve())
                except Exception:
                    return str(_state_dir() / e["file"])
    return None


def delete_entry(image_id: str) -> bool:
    """Remove one history entry and its image file. True when an entry was removed."""
    image_id = (image_id or "").strip()
    if not image_id:
        return False
    with _lock:
        doc = _load_index()
        entries = doc.get("entries") or []
        keep, removed = [], None
        for e in entries:
            if e.get("id") == image_id and removed is None:
                removed = e
            else:
                keep.append(e)
        if removed is None:
            return False
        doc["entries"] = keep
        _save_index(doc)
    rel = removed.get("file")
    if rel:
        try:
            (_state_dir() / rel).unlink()
        except OSError:
            pass
    return True


def clear_history() -> int:
    """Delete ALL history entries + image files. Returns the count removed."""
    with _lock:
        doc = _load_index()
        entries = doc.get("entries") or []
        base = _state_dir()
        for e in entries:
            rel = e.get("file")
            if rel:
                try:
                    (base / rel).unlink()
                except OSError:
                    pass
        n = len(entries)
        doc["entries"] = []
        _save_index(doc)
    return n


__all__ = [
    "record_image", "list_history", "read_image", "entry_path", "delete_entry", "clear_history",
]
